File: server/internal/handlers/experiment_fields.go
package handlers

import "fmt"

const (
	experimentNameMaxLength        = 64
	experimentKeyMaxLength         = 64
	experimentDescriptionMaxLength = 280
)

func validateExperimentName(name string) error {
	if name == "" {
		return fmt.Errorf("name is required")
	}
	if len(name) > experimentNameMaxLength {
		return fmt.Errorf("name must be 64 characters or fewer")
	}
	return nil
}

func validateExperimentKey(key string) error {
	if key == "" {
		return fmt.Errorf("key is required")
	}
	if len(key) > experimentKeyMaxLength {
		return fmt.Errorf("key must be 64 characters or fewer")
	}
	return nil
}

func validateExperimentDescription(description *string) error {
	if description != nil && len(*description) > experimentDescriptionMaxLength {
		return fmt.Errorf("description must be 280 characters or fewer")
	}
	return nil
}

File: client/src/lib/experimentFields.ts
export const EXPERIMENT_NAME_MAX_LENGTH = 64
export const EXPERIMENT_KEY_MAX_LENGTH = 64
export const EXPERIMENT_DESCRIPTION_MAX_LENGTH = 280

export function validateExperimentName(name: string): string | null {
  const trimmed = name.trim()
  if (!trimmed) return "Name is required."
  if (trimmed.length > EXPERIMENT_NAME_MAX_LENGTH) {
    return `Name must be ${EXPERIMENT_NAME_MAX_LENGTH} characters or fewer.`
  }
  return null
}

export function validateExperimentKey(key: string): string | null {
  const trimmed = key.trim()
  if (!trimmed) return "Key is required."
  if (trimmed.length > EXPERIMENT_KEY_MAX_LENGTH) {
    return `Key must be ${EXPERIMENT_KEY_MAX_LENGTH} characters or fewer.`
  }
  return null
}

export function validateExperimentDescription(description: string | null | undefined): string | null {
  if ((description ?? "").length > EXPERIMENT_DESCRIPTION_MAX_LENGTH) {
    return `Description must be ${EXPERIMENT_DESCRIPTION_MAX_LENGTH} characters or fewer.`
  }
  return null
}
